File: src/main/java/com/kasisoft/libs/common/KclException.java
package com.kasisoft.libs.common;

import jakarta.validation.constraints.*;

import java.util.function.*;

/**
 * Specialisation of the RuntimeException which is commonly used within this library.
 *
 * @author daniel.kasmeroglu@kasisoft.com
 */
public class KclException extends RuntimeException {

    private static final long serialVersionUID = -7284302896487719569L;

    public KclException() {
        super();
    }

    public KclException(@NotNull Exception ex) {
        super(ex);
    }

    public KclException(@NotBlank String msg) {
        super(msg);
    }

    public KclException(@NotNull Exception ex, @NotBlank String msg) {
        super(msg, ex);
    }

    public static <R> R execute(@NotNull Supplier<R> supplier, @NotBlank String msg) {
        try {
            return supplier.get();
        } catch (Exception ex) {
            throw KclException.wrap(ex, msg);
        }
    }

    @NotNull
    public static KclException wrap(@NotNull Exception ex) {
        if (ex instanceof KclException kex) {
            return kex;
        }
        return new KclException(ex);
    }

    @NotNull
    public static KclException wrap(@NotNull Exception ex, @NotBlank String msg) {
        if (ex instanceof KclException kex) {
            return kex;
        }
        return new KclException(ex, msg);
    }

    public static KclException unwrap(Exception ex) {
        if (ex != null) {
            if (ex instanceof KclException kex) {
                return kex;
            } else if (ex.getCause() instanceof Exception cex) {
                return unwrap(cex);
            }
        }
        return null;
    }

} /* ENDCLASS */

File: src/main/kotlin/com/kasisoft/libs/common/KclException.kt
package com.kasisoft.libs.common

import java.util.function.*

/**
 * Specialisation of the RuntimeException which is commonly used within this library.
 *
 * @author daniel.kasmeroglu@kasisoft.net
 */
open class KclException: RuntimeException {

    constructor() : super()

    constructor(ex: Exception) : super(ex)

    constructor(fmt: String, vararg args: Any) : super(formatString(fmt, *args))

    constructor(ex: Exception, fmt: String, vararg args: Any): super(formatString(fmt, *args), ex)

    companion object {

        private fun formatString(fmt: String, vararg args: Any): String =
            if (args.isNotEmpty()) {
                String.format(fmt, *args)
            } else {
                fmt
            }

        @JvmStatic
        fun wrap(ex: Exception): KclException =
            when (ex) {
                is KclException -> ex
                else -> KclException(ex)
            }

        @JvmStatic
        fun wrap(ex: Exception, fmt: String, vararg args: Any): KclException =
            when (ex) {
                is KclException -> ex
                else -> KclException(ex, fmt, *args)
            }


        @JvmStatic
        fun unwrap(ex: Exception?): KclException? =
            if  (ex is KclException) {
                ex
            } else if (ex!!.cause is Exception) {
                unwrap(ex.cause as Exception)
            } else {
                null
            }

        @JvmStatic
        @Deprecated("KASI")
        fun <R> execute(supplier: Supplier<R>, fmt: String, vararg args: Any): R {
            return try {
                supplier.get()
            } catch (ex: Exception) {
                throw wrap(ex, fmt, *args)
            }
        }

    } /* ENDOBJECT */

} /* ENDCLASS */